pkg/ddl/job_table.go

```diff
@@ -431,33 +374,77 @@ func (s *jobScheduler) checkAndUpdateClusterState(needUpdate bool) error {
 	return nil
 }
 
-func (s *jobScheduler) loadDDLJobAndRun(se *sess.Session) {
-	s.mu.RLock()
-	s.mu.hook.OnGetJobBefore()
-	s.mu.RUnlock()
+func (s *jobScheduler) loadAndDeliverJobs(se *sess.Session) error {
+	if s.generalDDLWorkerPool.available() == 0 && s.reorgWorkerPool.available() == 0 {
+		return nil
+	}
 
-	startTime := time.Now()
-	job, isReorg, err := s.getJob(se)
-	if job == nil || err != nil {
+	defer s.runningJobs.resetAllPending()
+
+	const getJobSQL = `select reorg, job_meta from mysql.tidb_ddl_job %s order by job_id`
+	var whereClause string
+	if ids := s.runningJobs.allIDs(); len(ids) > 0 {
+		whereClause = fmt.Sprintf("where job_id not in (%s)", ids)
+	}
+	sql := fmt.Sprintf(getJobSQL, whereClause)
+	rows, err := se.Execute(context.Background(), sql, "load_ddl_jobs")
+	if err != nil {
+		return errors.Trace(err)
+	}
+	for _, row := range rows {
+		reorgJob := row.GetInt64(0) == 1
+		targetPool := s.generalDDLWorkerPool
+		if reorgJob {
+			targetPool = s.reorgWorkerPool
+		}
+		jobBinary := row.GetBytes(1)
+
+		job := model.Job{}
+		err = job.Decode(jobBinary)
 		if err != nil {
-			logutil.DDLLogger().Warn("get job met error", zap.Duration("take time", time.Since(startTime)), zap.Error(err))
+			return errors.Trace(err)
 		}
-		return
-	}
-	s.mu.RLock()
-	s.mu.hook.OnGetJobAfter(job)
-	s.mu.RUnlock()
 
-	pool := s.generalDDLWorkerPool
-	if isReorg {
-		pool = s.reorgWorkerPool
-	}
-	wk, err := pool.get()
-	if err != nil || wk == nil {
-		logutil.DDLLogger().Debug(fmt.Sprintf("[ddl] no %v worker available now", pool.tp()), zap.Error(err))
-		return
+		involving := job.GetInvolvingSchemaInfo()
+		if targetPool.available() == 0 {
+			s.runningJobs.addPending(involving)
+			continue
```

Comment:
Why not block and wait?

Reply:
then DDL in another pool will wait much longer